halo2_gadgets/src/poseidon/pow5.rs

```diff
@@ -342,25 +342,24 @@ impl<
                 let initial_state = initial_state?;
 
                 // Load the input into this region.
-                let load_input_word = |i: usize| {
-                    let constraint_var = match input.0[i].clone() {
-                        Some(PaddedWord::Message(word)) => word,
-                        Some(PaddedWord::Padding(padding_value)) => region.assign_fixed(
-                            || format!("load pad_{}", i),
-                            config.rc_b[i],
-                            1,
-                            || Value::known(padding_value),
-                        )?,
-                        _ => panic!("Input is not padded"),
-                    };
-                    constraint_var
+                let load_input_word = |i: usize| match input.0[i].clone() {
+                    Some(PaddedWord::Message(word)) => word
                         .copy_advice(
                             || format!("load input_{}", i),
                             &mut region,
                             config.state[i],
                             1,
                         )
-                        .map(StateWord)
+                        .map(StateWord),
+                    Some(PaddedWord::Padding(padding_value)) => region
+                        .assign_advice_from_constant(
```

Comment:
This breaks compatibility with existing circuits, because the constants move from being assigned in the `rc_b` column to being assigned in the circuit-global constant space.

Instead, we should revert the change made in https://github.com/zcash/halo2/commit/65a89f099b2be9dbc4b8b13ba1b44b620c118568#diff-81e1b63d3841ada245a37c0d485d03d48fb2d477a946e5135ba4602eb0e7e6f4 and manually call `assign_advice` and `constrain_equal`, which would avoid this problem because we can pass `Value::known` into `assign_advice`.

Reply:
Separately, @therealyingtong and I discussed in Office Hours how to fix this in general. The core problem is that `AssignedCell::copy_advice` (correctly) propagates `Value::unknown`, but `assign_fixed` isn't propagating the fact that it was called with `Value::known` to the `AssignedCell` it returns.

The simplest fix would be to always evaluate `to` inside `assign_fixed`, but that would result in significant proving performance regressions for existing circuits (as this bug in question meant that you just couldn't write circuits that way, i.e. it failed safe). Ideally we could use memoization to remember `to` until the first time we need to evaluate it, then evaluate once, and remember the result going forward. We tried that approach and ran into both interior mutability needs, and object safety / lifetime issues related to the `RegionLayouter` trait.

Reply:
Wouldn't it be sufficient just to make sure that the `MockProver` fails in the same way as the real prover?